cmd/theatre-envconsul/acceptance/acceptance.go

```diff
@@ -60,13 +53,150 @@ spec:
+	SentinelSecretValue  = "eats-the-world"
+)
 
-	BeforeEach(func() {
-		config, err := clientcmd.BuildConfigFromFlags("", kubeConfigPath)
-		Expect(err).NotTo(HaveOccurred(), "failed to construct kubernetes config")
+type Runner struct{}
+
+// Prepare is used for configuring a Vault server in our acceptance tests to provide
+// Kubernetes authentication via service account.
+//
+// It does several things:
+//
+// - Mounts a kv2 secrets engine at secret/
+// - Creates a Kubernetes auth backend mounted at auth/kubernetes
+// - Configures the Kubernetes backend to authenticate against the currently detected
+//   Kubernetes API server (the current cluster, if run from within)
+// - For all successful Kubernetes logins, the user is assigned a token that maps to a
+//   cluster-reader policy, which permits reading of secrets from:
+//
+//   secret/data/kubernetes/{namespace}/{service-account-name}/*
+//
+func (r *Runner) Prepare(logger kitlog.Logger, config *rest.Config) error {
+	cfg := api.DefaultConfig()
+	cfg.Address = "http://localhost:8200"
+
+	transport := cfg.HttpClient.Transport.(*http.Transport)
+	transport.TLSClientConfig = nil
+
+	client, err := api.NewClient(cfg)
+	if err != nil {
+		return errors.Wrap(err, "failed to configure vault client")
+	}
+
+	client.SetToken("vault-token") // set in the acceptance overlay (config/overlays/acceptance)
+
+	mountPath := "secret"
+	mountOptions := &api.MountInput{
+		Type:        "kv",
+		Description: "Generic Vault kv mount",
+		Options: map[string]string{
+			"version": "2",
+		},
+	}
+
+	logger.Log("msg", "mounting secret engine", "path", mountPath, "options", mountOptions)
+	client.Sys().Unmount(mountPath)
+	if err := client.Sys().Mount(mountPath, mountOptions); err != nil {
+		return err
+	}
+
+	enableOptions := &api.EnableAuthOptions{
+		Type:        "kubernetes",
+		Description: "Permit authentication by Kubernetes service accounts",
+	}
+
+	logger.Log("msg", "enabling auth mount", "path", AuthBackendMountPath, "options", enableOptions)
+	client.Sys().DisableAuth(AuthBackendMountPath)
+	if err := client.Sys().EnableAuthWithOptions(AuthBackendMountPath, enableOptions); err != nil {
+		return err
+	}
+
+	var ca string
+
+	if string(config.CAData) == "" {
+		caBytes, err := ioutil.ReadFile(config.CAFile)
+		ca = string(caBytes)
+		if err != nil {
+			return errors.Wrap(err, "could not parse certificate for kubernetes")
+		}
+	} else {
+		ca = string(config.CAData)
+	}
+
+	// We'll be running the acceptance tests from outside the kubernetes cluster, where the
+	// API server will have an IP address that is relative to the host machine. When we're
+	// within the cluster, like Vault, we want to talk to kubernetes.default.svc to ensure
+	// we're tapping the host IP address.
+	backendConfigPath := fmt.Sprintf("auth/%s/config", AuthBackendMountPath)
+	backendConfig := map[string]interface{}{
+		"kubernetes_host":    "https://kubernetes.default.svc",
+		"kubernetes_ca_cert": string(ca),
```

Comment:
this seems like you're casting a string to a string. i would do `var ca []byte` and deal with bytes in the above section, only finally casting it to string here

Reply:
Yep true, the code is casting a few times here. I tend to try and avoid using byte slices where possible and to lean on strings, but that gets a bit messy in this instance.

Going to clean it up.